### PR TITLE
Refactoring trait pipeline (2/2)

### PR DESCRIPTION
**Pull request by eberrigan**

### CODE REVIEW

File: sleap_roots/bases.py [hunk 1]
```diff
@@ -114,110 +178,252 @@ def get_lateral_count(pts: np.ndarray):
...
+        )
     else:
         base_ys = _base_pts[:, 1]
         return base_ys
 
 
 def get_base_length(pts: np.ndarray, monocots: bool = False):
-    """Get lateral roots top and deepest bases distance in y axis.
+    """Get the distance on the y-axis from the top lateral root base to the
+        bottom lateral root base.
```
Comment (talmo) [inline sleap_roots/bases.py:277]: Should be 1 line

File: sleap_roots/bases.py [hunk 2]
```diff
@@ -114,110 +178,252 @@ def get_lateral_count(pts: np.ndarray):
...
 
     Args:
-        pts: lateral root landmarks as array of shape (instance, point, 2)
+        pts: root landmarks as array of shape (instances, point, 2)
+        or base_ys (instances).
```
Comment (talmo) [inline sleap_roots/bases.py:281]: indent

**Review by talmo — CHANGES_REQUESTED** [review]

File: sleap_roots/bases.py [hunk 3]
```diff
@@ -48,55 +47,120 @@ def get_root_lengths(pts: np.ndarray) -> np.ndarray:
     return total_lengths
 
 
-def get_root_lengths_max(lengths: np.ndarray) -> np.ndarray:
-    """Return maximum root length for all roots in a frame.
+def get_root_lengths_max(pts: np.ndarray) -> np.ndarray:
+    """
+    Return maximum root length for all roots in a frame.
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap_roots/bases.py]: Done

File: sleap_roots/bases.py [hunk 4]
```diff
@@ -48,55 +47,120 @@ def get_root_lengths(pts: np.ndarray) -> np.ndarray:
...
-def get_base_tip_dist(pts: np.ndarray) -> np.ndarray:
+def get_base_tip_dist(
+    base_pts: np.ndarray = None, tip_pts: np.ndarray = None, pts: np.ndarray = None
+) -> np.ndarray:
     """Return distance from root base to tip.
 
     Args:
+        base_pts: bases of roots (instances, 2)
+        tip_pts: tips of roots (instances, 2)
+        OR
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap_roots/bases.py]: Done

File: sleap_roots/bases.py [hunk 5]
```diff
@@ -48,55 +47,120 @@ def get_root_lengths(pts: np.ndarray) -> np.ndarray:
...
+    """Get gravitropism index based on primary_length_max and primary_base_tip_dist
+      or primary root points.
 
     Args:
-        pts: primary root landmarks as array of shape (1, node, 2)
+        primary_base_tip_dist: scalar of distance from base to tip of primary root
+        (longest primary root prediction used if there is more than one)
+        primary_length: scalar of length of primary root
+        (longest primary root prediction used if there is more than one)
+        OR
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap_roots/bases.py]: Done

File: sleap_roots/bases.py [hunk 6]
```diff
@@ -114,110 +178,252 @@ def get_lateral_count(pts: np.ndarray):
 
 
 def get_base_xs(pts: np.ndarray, monocots: bool = False) -> np.ndarray:
-    """Get x coordinations of base points.
+    """
+    Get x coordinates of the base of each lateral root.
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap_roots/bases.py]: Done

**Pull request merged**